script/tool/test/pubspec_check_command_test.dart

```diff
@@ -0,0 +1,334 @@
+${headerSection('plugin', isPlugin: true, includeIssueTracker: false)}
+${environmentSection()}
+${flutterSection(isPlugin: true)}
+${dependenciesSection()}
+${devDependenciesSection()}
+''');
+
+      final Future<List<String>> result =
+          runCapturingPrint(runner, <String>['pubspec-check']);
+
+      await expectLater(
+        result,
+        throwsA(const TypeMatcher<ToolExit>()),
+      );
+    });
+
+    test('fails when environment section is out of order', () async {
+      final Directory pluginDirectory = createFakePlugin('plugin',
+          withSingleExample: true, packagesDirectory: packagesDir);
+
+      pluginDirectory.childFile('pubspec.yaml').writeAsStringSync('''
+${headerSection('plugin', isPlugin: true)}
+${flutterSection(isPlugin: true)}
+${dependenciesSection()}
+${devDependenciesSection()}
+${environmentSection()}
+''');
+
+      final Future<List<String>> result =
+          runCapturingPrint(runner, <String>['pubspec-check']);
+
+      await expectLater(
+        result,
+        throwsA(const TypeMatcher<ToolExit>()),
+      );
+    });
+
+    test('fails when flutter section is out of order', () async {
+      final Directory pluginDirectory = createFakePlugin('plugin',
+          withSingleExample: true, packagesDirectory: packagesDir);
+
+      pluginDirectory.childFile('pubspec.yaml').writeAsStringSync('''
+${headerSection('plugin', isPlugin: true)}
+${flutterSection(isPlugin: true)}
+${environmentSection()}
+${dependenciesSection()}
+${devDependenciesSection()}
+''');
+
+      final Future<List<String>> result =
+          runCapturingPrint(runner, <String>['pubspec-check']);
+
+      await expectLater(
+        result,
+        throwsA(const TypeMatcher<ToolExit>()),
+      );
+    });
+
+    test('fails when dependencies section is out of order', () async {
+      final Directory pluginDirectory = createFakePlugin('plugin',
+          withSingleExample: true, packagesDirectory: packagesDir);
+
+      pluginDirectory.childFile('pubspec.yaml').writeAsStringSync('''
+${headerSection('plugin', isPlugin: true)}
+${environmentSection()}
+${flutterSection(isPlugin: true)}
+${devDependenciesSection()}
+${dependenciesSection()}
+''');
+
+      final Future<List<String>> result =
+          runCapturingPrint(runner, <String>['pubspec-check']);
+
+      await expectLater(
+        result,
+        throwsA(const TypeMatcher<ToolExit>()),
+      );
+    });
+
+    test('fails when devDependencies section is out of order', () async {
+      final Directory pluginDirectory = createFakePlugin('plugin',
+          withSingleExample: true, packagesDirectory: packagesDir);
+
+      pluginDirectory.childFile('pubspec.yaml').writeAsStringSync('''
+${headerSection('plugin', isPlugin: true)}
+${environmentSection()}
+${devDependenciesSection()}
+${flutterSection(isPlugin: true)}
+${dependenciesSection()}
+''');
+
+      final Future<List<String>> result =
+          runCapturingPrint(runner, <String>['pubspec-check']);
```

Comment:
Do we want to assert that the tool has exit code as 0 for success and not 0 for failures?

(What happens if somebody does `throw ToolExit(0);`? Would the process stop, but return a non-error exit code?)

Reply:
This is the pattern we have in all the existing tests. Rather than test for it everywhere, I've now added a failsafe in main where we handle ToolExit to use 255 if someone screws up and passes 0.

Reply:
Nice!